packages/@aws-cdk/core/lib/fs/ignore.ts

```diff
@@ -0,0 +1,183 @@
+import dockerIgnore, * as DockerIgnore from '@balena/dockerignore';
+import gitIgnore, * as GitIgnore from 'ignore';
+import * as minimatch from 'minimatch';
+import { CopyOptions, IgnoreMode } from './options';
+
+/**
+ * Represents file path ignoring behavior.
+ */
+export abstract class IgnoreStrategy {
+  /**
+   * Ignores file paths based on simple glob patterns.
+   *
+   * @returns `GlobIgnorePattern` associated with the given patterns.
+   * @param patterns
+   */
+  public static glob(patterns: string[]): GlobIgnoreStrategy {
+    return new GlobIgnoreStrategy(patterns);
+  }
+
+  /**
+   * Ignores file paths based on the [`.gitignore specification`](https://git-scm.com/docs/gitignore).
+   *
+   * @returns `GitIgnorePattern` associated with the given patterns.
+   * @param patterns
+   */
+  public static git(patterns: string[]): GitIgnoreStrategy {
+    return new GitIgnoreStrategy(patterns);
+  }
+
+  /**
+   * Ignores file paths based on the [`.dockerignore specification`](https://docs.docker.com/engine/reference/builder/#dockerignore-file).
+   *
+   * @returns `DockerIgnorePattern` associated with the given patterns.
+   * @param patterns
+   */
+  public static docker(patterns: string[]): DockerIgnoreStrategy {
+    return new DockerIgnoreStrategy(patterns);
+  }
+
+  /**
+   * Creates an IgnoreStrategy based on the `ignoreMode` and `exclude` in a `CopyOptions`.
+   *
+   * @returns `IgnoreStrategy` based on the `CopyOptions`
+   * @param options the `CopyOptions` to create the `IgnoreStrategy` from
+   */
+  public static fromCopyOptions(options: CopyOptions): IgnoreStrategy {
+    const ignoreMode = options.ignoreMode || IgnoreMode.GLOB;
+    const exclude = options.exclude || [];
+
+    switch (ignoreMode) {
+      case IgnoreMode.GLOB:
+        return this.glob(exclude);
+
+      case IgnoreMode.GIT:
+        return this.git(exclude);
+
+      case IgnoreMode.DOCKER:
+        return this.docker(exclude);
+    }
+  }
+
+  /**
+   * Adds another pattern.
+   * @params pattern the pattern to add
+   */
+  public abstract add(pattern: string): void;
+
+  /**
+   * Determines whether a given file path should be ignored or not.
+   *
+   * @param filePath file path to be assessed against the pattern
+   * @returns `true` if the file should be ignored
+   */
+  public abstract ignores(filePath: string): boolean;
```

Comment:
For all of these ignore functions, I'd like *very much* to be *very clear* about the expectations of absoluteness/relativeness of these file paths.

It seems like right now you're implicitly expecting all paths to be relative to the directory that contains the ignore file.

Please be very clear about that, documentation-wise, and add assertions/checks wherever it makes sense: reject absolute paths, etc.

Or, add in a way to normalize paths which can handle both absolute and relative. Or, expect absolute and relativize before checking. As long as the contract is very very clear.

Reply:
I agree. The root path against which paths passed to `ignores()` were being relativized was an implicit dependency of an `IgnoreStrategy`.

See fe58be11b2c25997bc5106e0d3df92b8c128b890

I made this an explicit dependency. All paths are expected to be absolute paths. This is documented and reflected in the variable naming. I also throw exceptions if relative paths are passed where absolute paths are expected (which is everywhere in this PR).